ml-agents/mlagents/trainers/optimizer/torch_optimizer.py

```diff
@@ -54,23 +55,71 @@ def create_reward_signals(self, reward_signal_configs):
             )
 
     def get_trajectory_value_estimates(
-        self, batch: AgentBuffer, next_obs: List[np.ndarray], done: bool
-    ) -> Tuple[Dict[str, np.ndarray], Dict[str, float]]:
+        self,
+        batch: AgentBuffer,
+        next_obs: List[np.ndarray],
+        done: bool,
+        agent_id: str = "",
+    ) -> Tuple[Dict[str, np.ndarray], Dict[str, float], Optional[AgentBufferField]]:
         n_obs = len(self.policy.behavior_spec.observation_specs)
-        current_obs = ObsUtil.from_buffer(batch, n_obs)
+
+        if agent_id in self.critic_memory_dict:
+            memory = self.critic_memory_dict[agent_id]
+        else:
+            memory = (
+                torch.zeros((1, 1, self.critic.memory_size))
+                if self.policy.use_recurrent
+                else None
+            )
+
+        # If we're using LSTM, we want to get all the intermediate memories.
+        all_next_memories: Optional[AgentBufferField] = None
+        if self.policy.use_recurrent:
+            resequenced_buffer = AgentBuffer()
+            all_next_memories = AgentBufferField()
+            # The 1st sequence are the ones that are padded. So if seq_len = 3 and
+            # trajectory is of length 10, the ist sequence is [pad,pad,obs].
+            # Compute the number of elements in this padded seq.
+            leftover = batch.num_experiences % self.policy.sequence_length
+            first_seq_len = self.policy.sequence_length if leftover == 0 else leftover
+            for _ in range(first_seq_len):
+                all_next_memories.append(memory.squeeze().detach().numpy())
+
+            batch.resequence_and_append(
+                resequenced_buffer, training_length=self.policy.sequence_length
+            )
+            reseq_obs = ObsUtil.from_buffer(resequenced_buffer, n_obs)
+            reseq_obs = [ModelUtils.list_to_tensor(obs) for obs in reseq_obs]
+            # By now, the buffer should be of length seq_len * num_seq, padded
+            _mem = memory
+            for seq_num in range(
+                resequenced_buffer.num_experiences // self.policy.sequence_length - 1
+            ):
+                seq_obs = []
+                for _obs in reseq_obs:
+                    start = seq_num * self.policy.sequence_length
+                    end = (seq_num + 1) * self.policy.sequence_length
+                    seq_obs.append(_obs[start:end])
+                _, next_seq_mem = self.critic.critic_pass(
```

Comment:
we can save the calls to get value estimates  later if we use the returned value here

Reply:
+1 it is a TODO